File: 59-SpiralMatrixIi/59-SpiralMatrixIi.py
# Last updated: 12/6/2025, 5:54:13 am
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        if n==0: return []
        if n==1: return [[1]]
        ret = [[0]*n for i in range(n)]
        x,y=0,0
        dir = 'r'
        for i in range(n*n):
            print(x,',',y,',',dir,',',i+1)
            ret[y][x] = i+1
            if dir == 'r':
                if x+1 >= n or ret[y][x+1] != 0:
                    dir = 'd'
                    y+=1
                    if y >= n: break
                else: x+=1
            elif dir == 'd':
                if y+1 >= n or ret[y+1][x] != 0:
                    dir = 'l'
                    x-=1
                    if x < 0: break
                else: y+=1
            elif dir == 'l':
                if x-1 < 0 or ret[y][x-1] != 0:
                    dir = 'u'
                    y-=1
                    if y < 0: break
                else: x-=1
            elif dir == 'u':
                if y-1 < 0 or ret[y-1][x] != 0:
                    dir = 'r'
                    x+=1
                    if x >= n: break
                else: y-=1
        return ret

File: 59-SpiralMatrixIi/59-SpiralMatrixIi.js
// Last updated: 12/6/2025, 5:54:18 am
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    if (n==0) return [];
    if (n==1) return [[1]];
	var ret=new Array(n).fill(0).map(()=> new Array(n).fill(0));
	var dir = 'x+'
	var x=0,y=0;
	var xp=1,xm=0,yp=0,ym=0;
	for (var i=0; i<n*n; i++){
		ret[y][x] = i+1;
		if (dir == 'x+'){
			if (x+1 >= n || ret[y][x+1] != 0){
				dir = 'y+';
				y++;
				if (y >= n) break;
			}
			else x++;
		}else if (dir == 'y+'){
			if (y+1 >= n || ret[y+1][x] != 0){
				dir = 'x-';
				x--;
				if (x < 0) break;
			}
			else y++;
		}else if (dir == 'x-'){
			if (x-1 < 0 || ret[y][x-1] != 0){
				dir = 'y-';
				y--;
				if (y < 0) break;
			}
			else x--;
		}else if (dir == 'y-'){
			if (y-1 < 0 || ret[y-1][x] != 0){
				dir = 'x+';
				x++;
				if (x >= n) break;
			}
			else y--;
		}
	}
	return ret;
};